File: 2015/16/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type Sue struct {
	number     int
	properties map[string]int
}

type GiftAnalyzer struct {
	sues []Sue
}

var MFCSAM_READING = map[string]int{
	"children":    3,
	"cats":        7,
	"samoyeds":    2,
	"pomeranians": 3,
	"akitas":      0,
	"vizslas":     0,
	"goldfish":    5,
	"trees":       3,
	"cars":        2,
	"perfumes":    1,
}

var GREATER_THAN_PROPS = []string{"cats", "trees"}

var LESS_THAN_PROPS = []string{"pomeranians", "goldfish"}

func NewGiftAnalyzer(path string) *GiftAnalyzer {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("Error reading input file: %v", err)
	}
	defer file.Close()

	var sues []Sue
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		fields := strings.Fields(line)
		if len(fields) < 8 {
			fmt.Fprintf(os.Stderr, "Skipping malformed line: %s\n", line)
			continue
		}

		numStr := strings.TrimSuffix(fields[1], ":")
		number, err := strconv.Atoi(numStr)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error parsing sue number in line: %s\n", line)
			continue
		}

		properties := make(map[string]int)

		key1 := strings.TrimSuffix(fields[2], ":")
		valStr1 := strings.TrimSuffix(fields[3], ",")
		val1, err := strconv.Atoi(valStr1)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error parsing property value in line: %s\n", line)
			continue
		}
		properties[key1] = val1

		key2 := strings.TrimSuffix(fields[4], ":")
		valStr2 := strings.TrimSuffix(fields[5], ",")
		val2, err := strconv.Atoi(valStr2)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error parsing property value in line: %s\n", line)
			continue
		}
		properties[key2] = val2

		key3 := strings.TrimSuffix(fields[6], ":")
		valStr3 := strings.TrimSuffix(fields[7], ",")
		val3, err := strconv.Atoi(valStr3)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error parsing property value in line: %s\n", line)
			continue
		}
		properties[key3] = val3

		sue := Sue{
			number:     number,
			properties: properties,
		}
		sues = append(sues, sue)
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("Error reading file: %v", err)
	}

	return &GiftAnalyzer{sues: sues}
}

func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

func (ga *GiftAnalyzer) isExactMatch(sue *Sue) bool {
	for prop, value := range sue.properties {
		if expected, ok := MFCSAM_READING[prop]; ok {
			if value != expected {
				return false
			}
		}
	}
	return true
}

func (ga *GiftAnalyzer) isRangeMatch(sue *Sue) bool {
	for prop, value := range sue.properties {
		expected, ok := MFCSAM_READING[prop]
		if !ok {
			continue
		}
		if contains(GREATER_THAN_PROPS, prop) {
			if value <= expected {
				return false
			}
		} else if contains(LESS_THAN_PROPS, prop) {
			if value >= expected {
				return false
			}

		} else {

			if value != expected {
				return false
			}
		}
	}
	return true
}

func (ga *GiftAnalyzer) findExactMatch() (int, bool) {
	for _, sue := range ga.sues {
		if ga.isExactMatch(&sue) {
			return sue.number, true
		}
	}
	return -1, false
}

func (ga *GiftAnalyzer) findRangeMatch() (int, bool) {
	for _, sue := range ga.sues {
		if ga.isRangeMatch(&sue) {
			return sue.number, true
		}
	}
	return -1, false
}

func main() {
	analyzer := NewGiftAnalyzer("../input.txt")
	sep := strings.Repeat("=", 20)

	if exactMatch, found := analyzer.findExactMatch(); found {
		fmt.Printf("%s Part 1 %s\nThe aunt sue that got you the gift is %d\n", sep, sep, exactMatch)
	} else {
		fmt.Println("No exact match found for Part 1")
	}

	if rangeMatch, found := analyzer.findRangeMatch(); found {
		fmt.Printf("%s Part 2 %s\nThe aunt sue that got you the gift is %d\n", sep, sep, rangeMatch)
	} else {
		fmt.Println("No range match found for Part 2")
	}
}

File: 2015/16/rust/src/main.rs
use std::{collections::HashMap, fs};

struct Sue {
    number: i32,
    properties: HashMap<String, i32>,
}

struct GiftAnalyzer {
    sues: Vec<Sue>,
}

impl GiftAnalyzer {
    const MFCSAM_READING: [(&'static str, i32); 10] = [
        ("children", 3),
        ("cats", 7),
        ("samoyeds", 2),
        ("pomeranians", 3),
        ("akitas", 0),
        ("vizslas", 0),
        ("goldfish", 5),
        ("trees", 3),
        ("cars", 2),
        ("perfumes", 1),
    ];

    const GREATER_THAN_PROPS: [&'static str; 2] = ["cats", "trees"];
    const LESS_THAN_PROPS: [&'static str; 2] = ["pomeranians", "goldfish"];

    pub fn new(path: &str) -> Self {
        let mut sues = Vec::new();

        let data = fs::read_to_string(path)
            .expect("Error reading input file")
            .trim()
            .lines()
            .map(|l| l.to_string())
            .collect::<Vec<_>>();

        for line in data {
            let sue: Vec<&str> = line.split_whitespace().collect();

            if sue.len() < 8 {
                eprintln!("Skipping malformed line: {}", line);
                continue;
            }

            let number = sue[1].trim_end_matches(':').parse::<i32>().unwrap();
            let item_1_nm = sue[2].trim_end_matches(':').to_string();
            let item_1_qtd = sue[3].trim_end_matches(',').parse::<i32>().unwrap();

            let item_2_nm = sue[4].trim_end_matches(':').to_string();
            let item_2_qtd = sue[5].trim_end_matches(',').parse::<i32>().unwrap();

            let item_3_nm = sue[6].trim_end_matches(':').to_string();
            let item_3_qtd = sue[7].trim_end_matches(',').parse::<i32>().unwrap();

            let mut properties = HashMap::new();
            properties.insert(item_1_nm, item_1_qtd);
            properties.insert(item_2_nm, item_2_qtd);
            properties.insert(item_3_nm, item_3_qtd);

            sues.push(Sue {
                number,
                properties,
            });
        }

        Self { sues }
    }

    fn is_exact_match(&self, sue: &Sue) -> bool {
        for (prop, &value) in sue.properties.iter() {
            if let Some(expected) = Self::MFCSAM_READING
                .iter()
                .find(|&&(k, _)| k == prop)
                .map(|&(_, v)| v)
            {
                if value != expected {
                    return false;
                }
            }
        }
        true
    }

    fn is_range_match(&self, sue: &Sue) -> bool {
        for (prop, value) in sue.properties.iter() {
           
            if let Some(expected) = Self::MFCSAM_READING
                .iter()
                .find(|&&(k, _)| k == prop)
                .map(|&(_, v)| v)
            {
                
                if Self::GREATER_THAN_PROPS.contains(&prop.as_str()) {
                    if value <= &expected {
                        return false;
                    }

                } else if Self::LESS_THAN_PROPS.contains(&prop.as_str()) {
                    if value >= &expected {
                        return false;
                    }
                    
                } else {
                
                    if value != &expected {
                        return false;
                    }
                }
            }
        }
        true
    }

    pub fn find_exact_match(&self) -> Option<i32> {
        for sue in &self.sues {
            if self.is_exact_match(sue) {
                return Some(sue.number);
            }
        }
        None
    }

    pub fn find_range_match(&self) -> Option<i32> {
        for sue in &self.sues {
            if self.is_range_match(sue) {
                return Some(sue.number);
            }
        }
        None
    }
}

fn main() {
    let analyzer = GiftAnalyzer::new("../input.txt");
    let sep = "=".repeat(20);
    let exact_match = analyzer.find_exact_match().unwrap();
    println!(
        "{} Part 1 {}\nThe aunt sue that got you the gift is {}",
        sep, sep, exact_match
    );

    let range_match = analyzer.find_range_match().unwrap();
    println!(
        "{} Part 2 {}\nThe aunt sue that got you the gift is {}",
        sep, sep, range_match
    );
}
